source/kernel/src/services/sdmmc.rs

```diff
@@ -0,0 +1,467 @@
+    /// 1-bit bus width, default after power-up or idle
+    Single,
+    /// 4-bit bus width, also called wide bus operation mode for SD cards
+    Quad,
+    /// 8-bit bus width, only available for MMC
+    Octo,
+}
+
+/// TODO
+#[derive(Debug, Eq, PartialEq)]
+pub enum HardwareOptions {
+    /// No change in configuration
+    None,
+    /// Switch the bus width
+    SetBusWidth(BusWidth),
+}
+
+/// TODO
+#[derive(Debug, Eq, PartialEq)]
+pub enum CommandKind {
+    /// Command without data transfer
+    Control,
+    /// Command for reading data, contains the number of bytes to read
+    Read(usize),
+    /// Command for writing data, contains the number of bytes to write
+    Write(usize),
+}
+
+/// TODO
+#[derive(Debug, Eq, PartialEq)]
+pub enum ResponseType {
+    /// No Response
+    None,
+    /// Response with 48-bit length
+    Short,
+    /// Response with 48-bit length + check *busy* after response
+    ShortWithBusySignal,
+    /// Response with 136-bit length
+    Long,
+}
+
+/// Response returned by the card, can be short or long, depending on command.
+#[must_use]
+pub enum Response {
+    /// The 32-bit value from the 48-bit response.
+    /// Potentially also includes the data buffer if this was a read command.
+    Short {
+        value: u32,
+        data: Option<FixedVec<u8>>,
+    },
+    /// The 128-bit value from the 136-bit response.
+    // TODO: make this `u128`?
+    Long([u32; 4]),
+}
+
+/// TODO
+#[derive(Debug, Eq, PartialEq)]
+pub enum Error {
+    /// TODO
+    Busy,
+    /// TODO
+    Response,
+    /// TODO
+    Data,
+    /// TODO
+    Timeout,
+    /// TODO
+    Other,
+}
+
+impl Default for Command {
+    fn default() -> Self {
+        Command {
+            index: 0,
+            argument: 0,
+            options: HardwareOptions::None,
+            kind: CommandKind::Control,
+            rsp_type: ResponseType::None,
+            rsp_crc: false,
+            buffer: None,
+        }
+    }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+// Client Definition
+////////////////////////////////////////////////////////////////////////////////
+
+/// A client for SD memory cards using the [`SdmmcService`].
+pub struct SdCardClient {
+    handle: KernelHandle<SdmmcService>,
+    reply: Reusable<Envelope<Result<Response, Error>>>,
+}
+
+/// The different types of cards
+#[derive(Debug, PartialEq)]
+pub enum CardType {
+    SD1,
+    SD2,
+    SDHC,
+}
+
+/// Card status in R1 response format
+#[derive(Debug, PartialEq)]
+pub struct CardStatus(u32);
+
+/// Card identification register in R2 response format
+///   Manufacturer ID       [127:120]
+///   OEM/Application ID    [119:104]
+///   Product name          [103:64]
+///   Product revision      [63:56]
+///   Product serial number [55:24]
+///   Reserved              [23:20]
+///   Manufacturing date    [19:8]
+///   CRC7 checksum         [7:1]
+///   Not used, always 1    [0:0]
+pub struct CardIdentification(u128);
```

Comment:
this could also potentially be a nice use for [`mycelium-bitfield`](https://docs.rs/mycelium-bitfield/latest/mycelium_bitfield/index.html) --- up to you though!